### PR TITLE
Fix Constraints endpoint giving 500 error

### LINKED ISSUE #1030: Improve failure behavior of constraints endpoint
## Description

This ticket lists many small (somewhat independent) problems with the response of the constraints API when subjected to POST requests containing invalid data.

## Caveats

- I tested the constraints API using the changes for #1026, which I'm presuming will soon be merged.
- I'm not mentioning some of the more obvious failure cases which also have the problems with non-conformant response (e.g. case where the request is missing the "columns" field), because I imagine we'll fix that at a lower level, so as to apply to all endpoints.

## Addition context

- #883
- [API Error Standards](https://wiki.mathesar.org/en/engineering/architecture/api-standards#errors)

## Response schema problems

In the following situations, the API correctly responds with a HTTP 4xx status code, but the response schema does not conform to our error standards.

- ~~Attempting to set a unique constraint for a column which contains duplicate values~~ (now fixed)
- ~~Attempting to set a unique constraint with a supplied constraint name equal to that of an existing table~~ (now fixed)

## Unhandled exception problems

In the following situations, the API responds with a HTTP 5xx status code when it should respond with a 4xx status code and a JSON response body conforming to our error standards.

- Invalid "type", e.g.:

    ```json
    {"type":"foo","columns":[1]}
    ```

- Attempting to set a unique constraint by supplying the id of a column that exists in a different table from the table whose id is supplied in the URL.

- An empty `columns` array, e.g.:

    ```json
    {"type":"unique","columns":[]}
    ```

- Non-existent column id, e.g.:

    ```json
    {"type":"unique","columns":[9999]}
    ```

### PR DESCRIPTION
**Pull request by Anish9901**
Fixes #1030

This PR Fixes many small unhandled exceptions and 500s caused by the constraints API endpoint.

**Technical details**

- An empty columns array:
```json
{"type":"unique","columns":[]}
```
- Response:
```json
{
    "code": 4005,
    "message": "Columns field cannot be empty.",
    "field": null,
    "detail": null
}
```
- _" Attempting to set a unique constraint by supplying the id of a column that exists in a different table from the table whose id is supplied in the URL"_ **and** Non-existent column id
```json
{"type":"unique","columns":[1, 10]}
```
**or**
```json
{"type":"unique","columns":[9999]}
```
- Response
```json
{
    "code": 4005,
    "message": "Column does not exist",
    "field": null,
    "detail": null
}
```

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by Anish9901** [comment]
@silentninja As there were many changes in PR #1375 in the constraints endpoint since I started working on this issue, I am yet to figure out a way to fix 500 error when a request with invalid type is made e.g. 
``` json
{"type":"foo","columns":[1]}
```

File: mathesar/models.py [hunk 1]
```diff
@@ -314,8 +316,12 @@ def delete_record(self, id_value):
         return delete_record(self._sa_table, self.schema._sa_engine, id_value)
 
     def add_constraint(self, constraint_type, columns, name=None):
+        if columns == []:
+            message = 'Columns field cannot be empty.'
+            raise base_api_exceptions.NotFoundAPIException(SyntaxError, message=message, status_code=status.HTTP_400_BAD_REQUEST)
         if constraint_type != constraint_utils.ConstraintType.UNIQUE.value:
-            raise ValueError('Only creating unique constraints is currently supported.')
+            message = 'Only creating unique constraints is currently supported.'
+            raise base_api_exceptions.ValueAPIException(ValueError, message=message, status_code=status.HTTP_400_BAD_REQUEST)
```
Comment (Anish9901) [inline mathesar/models.py:324]: @silentninja This is the change I originally made before #1375 got merged.

**Comment by silentninja** [comment]
> @silentninja As there were many changes in PR #1375 in the constraints endpoint since I started working on this issue, I am yet to figure out a way to fix 500 error when a request with invalid type is made e.g.
> 
> ```json
> {"type":"foo","columns":[1]}
> ```
> 

We need to raise a Validation error or have a default serializer in case of unknown choice, each has its own set of problems

- Default Serializer would allow invalid options too
- Throwing up validation error would mean we need to know all the possible values of the database.

I need to think about this, we might have to handle it on a case by case basis

**Comment by Anish9901** [comment]
> I need to think about this, we might have to handle it on a case by case basis

Sure I'll wait until the PR gets reviewed.

File: mathesar/models.py (hunk 1 above)
Comment (Anish9901) [reply mathesar/models.py]: @silentninja Passing an invalid type in the request still results in an Internal Server Error.

**Comment by Anish9901** [comment]
@silentninja I think I messed up something while resolving some merge conflicts. I'll create a new PR.

**Comment by Anish9901** [comment]
New PR #1406

**Pull request closed without merging**